File: framework/languages/dotnet/src/Zlink.HttpClient/Runtime/RequestPerformer.cs
/* SPDX-License-Identifier: Apache-2.0 */

using System.Net;
using System.Text;
using SystemHttpClient = System.Net.Http.HttpClient;

namespace Zlink.HttpClient.Runtime;

/// <summary>
///     Performs a single logical request, driving the wrapper-owned redirect loop, cookie jar
///     integration, and same-origin <c>Authorization</c> scrubbing. Redirect/URL rules live in
///     <see cref="HttpRedirectPolicy" /> and response decoding in <see cref="ResponseBodyReader" />.
///     Mirrors the C++ <c>request_performer.cpp</c>. Native auto-redirect/auto-decompress/auto-cookie are
///     disabled on the handler (see <see cref="HttpClientRuntime" />) so these semantics match the ZLink
///     contract.
/// </summary>
internal sealed class RequestPerformer(
    HttpClientOptions options,
    CookieJar cookieJar,
    SystemHttpClient httpClient)
{
    private readonly ResponseBodyReader _bodyReader = new(options);

    public async ValueTask<RawHttpResponse> PerformAsync(HttpRequestSpec request, CancellationToken cancellationToken)
    {
        var baseUri = new Uri(options.BaseUrl);
        // GetLeftPart(Authority) keeps IPv6 literals bracketed (e.g. http://[::1]:8080), which string
        // interpolation of Host+Port would break.
        var origin = new Uri(baseUri.GetLeftPart(UriPartial.Authority));
        var combinedTarget = HttpRedirectPolicy.MakeTarget(baseUri.AbsolutePath, request.Target);
        var current = new Uri(origin, combinedTarget);
        var authorizationAllowed = HttpRedirectPolicy.SameOrigin(origin, current);
        var method = request.Method;
        var body = request.Body;
        // A streamed body provider cannot be rewound, so it is dropped once a redirect is followed
        // (kept here, not in HttpRedirectPolicy, so all body-source ownership stays in one place).
        var bodyProvider = request.BodyProvider;
        var redirectsLeft = options.FollowRedirects;

        while (true)
        {
            using var message = BuildMessage(
                method,
                body,
                bodyProvider,
                request.Headers,
                current,
                authorizationAllowed);
            var completion = request.Sink is null
                ? HttpCompletionOption.ResponseContentRead
                : HttpCompletionOption.ResponseHeadersRead;
            using var response = await httpClient
                .SendAsync(message, completion, cancellationToken)
                .ConfigureAwait(false);

            var status = (int)response.StatusCode;
            if (options.Cookies && response.Headers.TryGetValues("Set-Cookie", out var setCookies))
                foreach (var setCookie in setCookies)
                    cookieJar.Store(current.Host, setCookie);

            var location = response.Headers.TryGetValues("Location", out var locations)
                ? locations.FirstOrDefault()
                : null;
            if (options.FollowRedirects > 0 && HttpRedirectPolicy.IsRedirectStatus(status) &&
                !string.IsNullOrEmpty(location))
            {
                if (redirectsLeft == 0)
                    throw new ZLinkFrameworkException(
                        ZLinkFrameworkErrorKind.ProtocolError,
                        "HTTP request exceeded the redirect limit");

                --redirectsLeft;
                (method, body) = HttpRedirectPolicy.RewriteForRedirect(status, method, body);
                bodyProvider = null; // consumed; never replay a non-rewindable stream on a redirect
                var next = HttpRedirectPolicy.ResolveLocation(current, location!);
                authorizationAllowed &= HttpRedirectPolicy.SameOrigin(current, next);
                current = next;
                continue;
            }

            var headers = ResponseBodyReader.CollectHeaders(response);
            if (request.Sink is not null)
            {
                await _bodyReader.StreamToSinkAsync(response, request.Sink, cancellationToken).ConfigureAwait(false);
                return new RawHttpResponse
                {
                    Status = status,
                    Headers = headers,
                    BodyBytes = []
                };
            }

            var bytes = await _bodyReader.ReadBufferedAsync(response, cancellationToken).ConfigureAwait(false);
            if (options.Compression) (bytes, headers) = _bodyReader.Decompress(bytes, headers);

            return new RawHttpResponse
            {
                Status = status,
                Headers = headers,
                BodyBytes = bytes
            };
        }
    }

    private HttpRequestMessage BuildMessage(
        ZLinkHttpMethod method,
        byte[]? body,
        Func<byte[]?>? bodyProvider,
        IReadOnlyDictionary<string, string> requestHeaders,
        Uri target,
        bool keepAuthorization)
    {
        var message = new HttpRequestMessage(ToHttpMethod(method), target)
        {
            Version = HttpVersion.Version11
        };

        var contentType = HttpHeaderLookup.Find(requestHeaders, "content-type");
        var streaming = bodyProvider is not null;
        HttpContent? content = null;
        if (bodyProvider is not null)
        {
            content = new StreamContent(new ProviderReadStream(bodyProvider));
            content.Headers.TryAddWithoutValidation("Content-Type", contentType ?? "application/octet-stream");
        }
        else if (body is not null)
        {
            content = new ByteArrayContent(body);
            if (contentType is not null) content.Headers.TryAddWithoutValidation("Content-Type", contentType);
        }

        message.Content = content;
        if (streaming) message.Headers.TransferEncodingChunked = true;

        // HttpListener (and many servers) do not emit 100-Continue; waiting for it would deadlock
        // a streamed/chunked upload. We never use the Expect/continue handshake.
        message.Headers.ExpectContinue = false;
        message.Headers.TryAddWithoutValidation("User-Agent", HttpClientVersion.UserAgent);
        message.Headers.TryAddWithoutValidation("Accept", "application/json");

        if (options.Compression) message.Headers.TryAddWithoutValidation("Accept-Encoding", "gzip, deflate");

        ApplyHeaders(message, options.Headers, keepAuthorization);
        ApplyHeaders(message, requestHeaders, keepAuthorization);

        if (options.Cookies)
        {
            var path = HttpRedirectPolicy.PathOf(target);
            var cookieHeader = cookieJar.HeaderFor(target.Host, path, target.Scheme == "https");
            if (cookieHeader.Length > 0) message.Headers.TryAddWithoutValidation("Cookie", cookieHeader);
        }

        return message;
    }

    private static void ApplyHeaders(
        HttpRequestMessage message,
        IReadOnlyDictionary<string, string> headers,
        bool keepAuthorization)
    {
        foreach (var (name, value) in headers)
        {
            if (name.Equals("content-type",
                    StringComparison.OrdinalIgnoreCase)) continue; // routed to the content above

            if (!keepAuthorization && name.Equals("authorization", StringComparison.OrdinalIgnoreCase)) continue;

            message.Headers.Remove(name);
            message.Headers.TryAddWithoutValidation(name, value);
        }
    }

    private static HttpMethod ToHttpMethod(ZLinkHttpMethod method)
    {
        return method switch
        {
            ZLinkHttpMethod.Get => HttpMethod.Get,
            ZLinkHttpMethod.Post => HttpMethod.Post,
            ZLinkHttpMethod.Put => HttpMethod.Put,
            ZLinkHttpMethod.Delete => HttpMethod.Delete,
            ZLinkHttpMethod.Patch => HttpMethod.Patch,
            ZLinkHttpMethod.Head => HttpMethod.Head,
            ZLinkHttpMethod.Options => HttpMethod.Options,
            _ => HttpMethod.Get
        };
    }

}

File: framework/languages/node/packages/http-client/src/runtime/request-performer.ts
/* SPDX-License-Identifier: Apache-2.0 */

import { Readable } from 'node:stream';
import { request, type Dispatcher } from 'undici';
import { ZLinkFrameworkException, ZLinkFrameworkErrorKind } from '@zlink-systems/framework';
import type { BodyChunkProvider, DownloadSink, ZLinkHttpMethod } from '../types';
import type { HttpClientOptions } from './options';
import { CookieJar } from './cookie-jar';
import { isRedirectStatus, makeTarget, resolveLocation, rewriteForRedirect } from './redirect-policy';
import { ResponseBodyReader } from './response-body-reader';
import { httpClientUserAgent } from './version';

export interface HttpRequestSpec {
  readonly method: ZLinkHttpMethod;
  readonly target: string;
  readonly body?: string;
  readonly bodyProvider?: BodyChunkProvider;
  readonly headers: Readonly<Record<string, string>>;
  readonly timeoutMs?: number;
  readonly sink?: DownloadSink;
}

export interface RawResult {
  readonly status: number;
  readonly headers: Record<string, string>;
  readonly body: string;
}

/**
 * Performs a single logical request, driving the wrapper-owned redirect loop, cookie jar
 * integration, and same-origin `Authorization` scrubbing. Redirect/URL rules live in
 * `redirect-policy` and response decoding in {@link ResponseBodyReader}. Mirrors the C++
 * `request_performer.cpp`. undici does not auto-redirect, auto-decompress, or keep cookies, so these
 * semantics match the ZLink contract.
 */
export class RequestPerformer {
  private readonly bodyReader: ResponseBodyReader;

  constructor(
    private readonly options: HttpClientOptions,
    private readonly cookieJar: CookieJar,
    private readonly dispatcher: Dispatcher | undefined,
  ) {
    this.bodyReader = new ResponseBodyReader(options);
  }

  async perform(spec: HttpRequestSpec, signal: AbortSignal): Promise<RawResult> {
    const baseUri = new URL(this.options.baseUrl);
    const origin = baseUri.origin;
    let current = new URL(origin + makeTarget(baseUri.pathname, spec.target));
    let method = spec.method;
    let body = spec.body;
    // A streamed body provider cannot be rewound, so it is dropped once a redirect is followed.
    let bodyProvider = spec.bodyProvider;
    let redirectsLeft = this.options.followRedirects;

    for (;;) {
      const keepAuthorization = current.origin === origin;
      const hasBody = body !== undefined || bodyProvider !== undefined;
      const headers = this.buildHeaders(spec, current, keepAuthorization, hasBody);
      const response = await request(current.href, {
        method,
        headers,
        body: this.buildBody(body, bodyProvider),
        dispatcher: this.dispatcher,
        maxRedirections: 0,
        signal,
      });

      const status = response.statusCode;
      if (this.options.cookies) {
        const setCookie = response.headers['set-cookie'];
        const values = Array.isArray(setCookie) ? setCookie : setCookie === undefined ? [] : [setCookie];
        for (const value of values) {
          this.cookieJar.store(current.hostname, value);
        }
      }

      const location = headerValue(response.headers, 'location');
      if (this.options.followRedirects > 0 && isRedirectStatus(status) && location !== undefined && location.length > 0) {
        if (redirectsLeft === 0) {
          await drain(response.body);
          throw requestError('HTTP request exceeded the redirect limit');
        }
        redirectsLeft--;
        ({ method, body } = rewriteForRedirect(status, method, body));
        bodyProvider = undefined; // consumed; never replay a non-rewindable stream on a redirect
        await drain(response.body);
        current = resolveLocation(current, location);
        continue;
      }

      const collectedHeaders = ResponseBodyReader.collectHeaders(response.headers);
      if (spec.sink !== undefined) {
        await this.bodyReader.streamToSink(response.body, spec.sink);
        return { status, headers: collectedHeaders, body: '' };
      }

      let bytes = await this.bodyReader.readBuffered(response.body);
      let finalHeaders = collectedHeaders;
      if (this.options.compression) {
        const result = await this.bodyReader.decompress(bytes, finalHeaders);
        bytes = result.body;
        finalHeaders = result.headers;
      }
      // Decode lazily: the UTF-8 conversion of a large body is a main-thread cost that
      // callers who never read the text body (binary consumers) should not pay.
      let text: string | undefined;
      return {
        status,
        headers: finalHeaders,
        get body(): string {
          text ??= bytes.toString('utf8');
          return text;
        },
      };
    }
  }

  private buildHeaders(
    spec: HttpRequestSpec,
    target: URL,
    keepAuthorization: boolean,
    hasBody: boolean,
  ): Record<string, string> {
    const headers: Record<string, string> = {
      'user-agent': httpClientUserAgent,
      accept: 'application/json',
    };
    if (this.options.compression) {
      headers['accept-encoding'] = 'gzip, deflate';
    }
    // Proxy authentication is carried by the ProxyAgent dispatcher (see runtime.ts), not a header,
    // so it is not leaked to the target over a CONNECT tunnel.

    applyHeaders(headers, this.options.headers, keepAuthorization);
    applyHeaders(headers, spec.headers, keepAuthorization);

    if (!hasBody) {
      // A body source dropped by a redirect must not leave stale content-type metadata behind.
      delete headers['content-type'];
    }

    if (this.options.cookies) {
      const cookieHeader = this.cookieJar.headerFor(target.hostname, target.pathname, target.protocol === 'https:');
      if (cookieHeader.length > 0) {
        headers['cookie'] = cookieHeader;
      }
    }
    return headers;
  }

  private buildBody(
    body: string | undefined,
    bodyProvider: BodyChunkProvider | undefined,
  ): string | Readable | undefined {
    if (bodyProvider !== undefined) {
      const provider = bodyProvider;
      return Readable.from(
        (function* () {
          for (let chunk = provider(); chunk !== null; chunk = provider()) {
            if (chunk.length > 0) {
              yield chunk;
            }
          }
        })(),
      );
    }
    return body;
  }
}

function applyHeaders(
  target: Record<string, string>,
  headers: Readonly<Record<string, string>>,
  keepAuthorization: boolean,
): void {
  for (const [name, value] of Object.entries(headers)) {
    const lower = name.toLowerCase();
    if (!keepAuthorization && lower === 'authorization') {
      continue;
    }
    target[lower] = value;
  }
}

function headerValue(
  headers: Record<string, string | string[] | undefined>,
  name: string,
): string | undefined {
  const value = headers[name];
  if (value === undefined) {
    return undefined;
  }
  return Array.isArray(value) ? value[0] : value;
}

async function drain(stream: Readable): Promise<void> {
  for await (const _chunk of stream) {
    // discard
  }
}

function requestError(message: string): ZLinkFrameworkException {
  return new ZLinkFrameworkException(ZLinkFrameworkErrorKind.Unavailable, message);
}
